app/lifecycle/hook.go

```diff
@@ -0,0 +1,201 @@
+
+func (fn HookFuncErr) Call(context.Context) error {
+	return fn()
+}
+
+// HookFuncCtx wraps a context (no error) hook function as a IHookFunc.
+type HookFuncCtx func(ctx context.Context)
+
+func (fn HookFuncCtx) Call(ctx context.Context) error {
+	fn(ctx)
+	return nil
+}
+
+// HookStartType defines the type of start hook.
+type HookStartType int
+
+const (
+	// AsyncAppCtx defines a start hook that will be called asynchronously (non-blocking)
+	// with the application context. Using the application usually results in hard shutdown.
+	AsyncAppCtx HookStartType = iota + 1
+
+	// SyncBackground defines a start hook that wil be called synchronously (blocking)
+	// with a fresh background context. Processes that support graceful shutdown can
+	// associate this with a call to RegisterStop.
+	SyncBackground
+
+	// AsyncBackground defines a start hook that wil be called asynchronously (non-blocking)
+	// with a fresh background context. Processes that support graceful shutdown can
+	// associate this with a call to RegisterStop.
+	AsyncBackground
+)
+
+// hook represents a life cycle hook; either a start or a stop.
+type hook struct {
+	// Order defines the order in which hooks are called.
+	Order int
+	// Label is a text label for errors and logging.
+	Label string
+	// StartType defines whether the start type is (a)synchronous and which context to use.
+	StartType HookStartType
+	// Func is the hook function.
+	Func IHookFunc
+}
+
+// runHooks starts and stops all the provided hooks.
+func runHooks(appCtx context.Context, startHooks []hook, stopHooks []hook) error {
+	// Collect any first error, to return at the end.
+	firstErr := make(chan error, 1)
+	cacheErr := func(err error) {
+		select {
+		case firstErr <- err:
+		default:
+			// Some other error already first.
+		}
+	}
+
+	// startAppCtx is cancelled when app is shutdown or when starting a hook fails.
+	startAppCtx, cancel := context.WithCancel(appCtx)
+	defer cancel()
+
+	// backgroundCtx is never closed, it is provided to StartSyncBackground and StartAsyncBackground hooks,
+	// they are explicitly stopped.
+	backgroundCtx := log.WithTopic(context.Background(), "app-start")
+
+	err := startAllHooks(startAppCtx, backgroundCtx, startHooks, cancel, cacheErr)
+	if err != nil {
+		return err
+	}
+
+	// Wait for shutdown or hook start failure
+	<-startAppCtx.Done()
+
+	if appCtx.Err() != nil {
+		log.Info(appCtx, "Shutdown signal detected")
+	}
+
+	// stopCtx is a fresh context allowing 10s for shutdown.
+	stopCtx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	stopCtx = log.WithTopic(stopCtx, "app-stop")
+	log.Info(stopCtx, "Shutting down gracefully")
+
+	stopAllHooks(stopCtx, stopHooks, cancel, cacheErr)
+
+	cacheErr(nil) // Ensure there is something in firstErr.
+
+	return <-firstErr
+}
+
+// startHook starts a hook, block until it returns.
+func startHook(ctx context.Context, hook hook, cancel context.CancelFunc, cacheErr func(err error)) {
```

Comment:
I think we can try to do normal function that returns an error, then you can do context and error handling in the parent function.

Reply:
I don't understand, are you talking about cacheErr?

Reply:
ok yeah, I see that is not possible. But the problem is that even though you broke up the big function into smaller functions, you haven't achieved any form of decoupling. The inner functions references all the state they referenced previously. 

So technically decreased the lines of code per function, but I think this is harder to read and understand.

I had a look and do not see an easy way to simplify and decouple the code into proper smaller decoupled functions.

Reply:
I don't that because there are no additional advantages to these beyond the splitting the functions into smaller functions we should not take this advantage. This code is incrementally better that what we have.